src/Features/Core/Portable/ExtractClass/ExtractClassWithDialogCodeAction.cs

```diff
@@ -0,0 +1,260 @@
+
+            var remainingResults = new List<ExtractClassMemberAnalysisResult>(memberAnalysisResults);
+
+            foreach (var documentId in symbolMapping.DocumentIds)
+            {
+                if (remainingResults.Count == 0)
+                {
+                    // All symbols have been taken care of
+                    break;
+                }
+
+                var document = solution.GetRequiredDocument(documentId);
+
+                var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+                var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+                var typeDeclaration = root.GetAnnotatedNodes(symbolMapping.TypeNodeAnnotation).SingleOrDefault();
+                if (typeDeclaration == null)
+                {
+                    continue;
+                }
+
+                foreach (var memberAnalysis in remainingResults.ToArray())
+                {
+                    var annotation = symbolMapping.SymbolToDeclarationAnnotationMap[memberAnalysis.Member];
+
+                    var nodeOrToken = typeDeclaration.GetAnnotatedNodesAndTokens(annotation).SingleOrDefault();
+                    var node = nodeOrToken.IsNode
+                        ? nodeOrToken.AsNode()
+                        : nodeOrToken.AsToken().Parent;
+
+                    if (node is null)
+                    {
+                        continue;
+                    }
+
+                    var currentSymbol = semanticModel.GetDeclaredSymbol(node, cancellationToken);
+                    if (currentSymbol is null)
+                    {
+                        continue;
+                    }
+
+                    pullMembersBuilder.Add((currentSymbol, memberAnalysis.MakeAbstract));
+                    remainingResults.Remove(memberAnalysis);
+                }
+            }
+
+            Contract.ThrowIfFalse(remainingResults.Count == 0);
+
+            var pullMemberUpOptions = PullMembersUpOptionsBuilder.BuildPullMembersUpOptions(newType, pullMembersBuilder.ToImmutable());
+            var updatedOriginalDocument = solution.GetRequiredDocument(_document.Id);
+
+            return await MembersPuller.PullMembersUpAsync(updatedOriginalDocument, pullMemberUpOptions, cancellationToken).ConfigureAwait(false);
+        }
+
+        private static async Task<INamedTypeSymbol> GetNewTypeSymbolAsync(Document document, INamedTypeSymbol typeToBeLookedFor, CancellationToken cancellationToken)
+        {
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            return (INamedTypeSymbol)semanticModel.Compilation.GetSymbolsWithName(typeToBeLookedFor.Name, SymbolFilter.Type, cancellationToken).Single();
```

Comment:
1. How do you know `Single` will work?
2. can we not place an annotation on the newly generated node so we can just find it post generation?  

I really don't like GetSymbolsWithName here... it's highly highly sketchy.

Reply:
Changed `ExtractTypeHelpers.XXX` to annotate the new symbol that is extracted so we can look it up that way.